diff -urNp old/converter/ppm/Makefile new/converter/ppm/Makefile
--- old/converter/ppm/Makefile	2017-09-05 13:40:36.789469114 +0200
+++ new/converter/ppm/Makefile	2017-09-05 13:42:55.974616951 +0200
@@ -7,7 +7,7 @@ VPATH=.:$(SRCDIR)/$(SUBDIR)
 
 include $(BUILDDIR)/config.mk
 
-SUBDIRS = hpcdtoppm ppmtompeg
+SUBDIRS = hpcdtoppm
 
 PORTBINARIES =	411toppm eyuvtoppm gouldtoppm ilbmtoppm imgtoppm \
 		leaftoppm mtvtoppm neotoppm \
